docs/rfcs/blocks-in-widget-areas.md

```diff
@@ -0,0 +1,535 @@
+
+[sidebars_widgets]: https://developer.wordpress.org/reference/hooks/sidebars_widgets/
+
+### `dynamic_sidebar()`
+
+Usage of the [`dynamic_sidebar()`][dynamic_sidebar] function remains as-is, though it will now render any blocks in the specified widget area.
+
+There are some nuances that surround how `dynamic_sidebar()` handles the `'before_widget'`, `'after_widget'`, `'before_title'` and `'after_title'` attributes that are associated with the widget area.
+
+The `'before_title'` and `'after_title'` attributes are rendered:
+
+- Before/after the title of a `core/legacy-widget` block is rendered.
+- Never, when a non-`core/legacy-widget` block is rendered.
+
+The `'before_widget'` and `'after_widget'` attributes are rendered:
+
+- Before/after a `core/legacy-widget` block is rendered.
+- Before/after a _sequence_ of non-`core/legacy-widget` blocks are rendered.
+
+This is done so that all blocks aren't wrapped with `'before_widget'` and `'after_widget'` markup. Instead, **sequences** of blocks are combined together _and then_ wrapped with `'before_widget'` and `'after_widget'` markup. This makes blocks render on old themes as if they were all together in a single widget.
+
+To illustrate, let's say that the `'footer'` widget area is registered so that titles are wrapped in `<h2>` elements and widgets in `<section>` elements.
+
+```php
+register_sidebar( array( 
+	'id'            => 'footer',
+	...
+	'before_widget' => '<section>',
+	'after_widget'  => '</section>',
+	'before_title'  => '<h2>',
+	'after_title'   => '</h2>',
+) );
+```
+
+The `'footer'` widget area has been migrated.
+
+```php
+array(
+	'footer' => 123,
+	'array_version' => 4,
+)
+```
+
+The referenced `wp_area` contains some blocks and a legacy widget in its `post_content`.
+
+```html
+<!-- wp:paragraph --><p>Hello there!</p><!-- /wp:paragraph -->
+<!-- wp:legacy-widget {"identifier":"search-2","instance":{"title":"Search"}} /-->
+<!-- wp:separator --><hr class="wp-block-separator"/><!-- /wp:separator -->
+<!-- wp:paragraph --><p>Welcome to my cool website!</p><!-- /wp:paragraph -->
+```
+
+Now, let's call `dynamic_sidebar( 'footer' )`.
+
+```html
+<section>
+	<p>Hello there!</p>
+</section>
+<section>
+	<h2>Search</h2>
+	<form role="search" method="get" class="search-form" action="http://wordpress.test/">
+		...
+	</form>
+</section>
+<section>
+	<hr class="wp-block-seperator" />
+	<p>Welcome to my cool website!</p>
+</section>
+```
+
+[dynamic_sidebar]: https://developer.wordpress.org/reference/functions/dynamic_sidebar/
+[register_sidebar]: https://developer.wordpress.org/reference/functions/register_sidebar/
+
+## Performance
+
+`wp_area` posts will be pre-emptively fetched when rendering a page. That is, all of the posts referenced in `'sidebar_widgets'` will be fetched using a single `SELECT ... FROM wp_posts WHERE ID in ?` query. This ensures that we do not introduce one extra query per widget area.
```

Comment:
Are there performance issues around storing many widget / blocks in one post like this? How about each widget is a single post with a single block and they are groups with a taxonomy?

Reply:
Or a widget area is a wp_area post and the widgets are all reusable blocks?

Reply:
I think these two things `wp_area` and `wp_block` (reusable blocks) are two separate things with two different use-cases. They share technical aspects though, they have the exact same type of content (list of blocks), they can be both considered "templates" (in the broad sense, not WP one)

Reply:
Okay, agree regarding reuseable blocks. But what about my point on lots of blocks in one post? Some sidebars have tonnes of blocks. Worried about this scaling.

Reply:
How is this different from a regular post/page containing tonnes of blocks?

Reply:
> How is this different from a regular post/page containing tonnes of blocks?

Yeah, I don't like them either, they don't scale great, but we couldn't develop it different without breaking BC. With this, we have more options.

Reply:
I would personally like to see widgets stored individually in their own posts. It makes sense for multiple blocks to be stored together when writing a post or page, but imho widgets are distinct pieces of content and should continue to be stored that way. The current method of storing widgets is not ideal, but it does at least allow us to fairly easily extract or save a single widget instance. That doesn't seem so possible if all widgets within a sidebar as lumped together.

Storing them individually would also open up the ability to add metadata to a widget, something that until now has required 'hacking' the options table.

Reply:
Considering widgets will move at some points to blocks entirely @megamenu would saving each widget individually be still needed?

Reply:
@draganescu I understand the core widgets all have their own block equivalent, and at some time maybe those core widgets will be gone, but I'm assuming there will always be some form of support for custom "legacy" widgets?

Reply:
Perhaps @youknowriad can speak better to this, but I'm drawn to storing multiple blocks per widget area because it fits nicely with my mental model of how WordPress will look in the future which is that one's site will consist entirely of _block areas_. In this sense, the similarities between `post_content`, `wp_block` and `wp_area` are intentional.

```
+------------------------------------------------------------------------------+
|                                                                              |
|                               "header" wp_area                               |
|                                                                              |
+------------------------------------------------------------------------------+
+---------------------------------------------------+  +-----------------------+
|                                                   |  |                       |
|                                                   |  |                       |
|                                                   |  |                       |
|                                                   |  |                       |
|                                                   |  |                       |
|                                                   |  |                       |
|                                                   |  |                       |
|                   post_content                    |  |                       |
|                                                   |  |                       |
|                                                   |  |   "sidebar" wp_area   |
|                                                   |  |                       |
|                                                   |  |                       |
|                                                   |  |                       |
|                                                   |  |                       |
|                                                   |  |                       |
|  +---------------------------------------------+  |  |                       |
|  |                                             |  |  |                       |
|  |             "author bio" wp_block           |  |  |                       |
|  |                                             |  |  |                       |
|  +---------------------------------------------+  |  |                       |
|                                                   |  |                       |
+---------------------------------------------------+  +-----------------------+
+------------------------------------------------------------------------------+
|                                                                              |
|                                                                              |
|                                                                              |
|                               "footer" wp_area                               |
|                                                                              |
|                                                                              |
|                                                                              |
+------------------------------------------------------------------------------+
```